LinkedBlockingQueue dealloc chain, and mutex memory cycle

kpgalligan opened the pull request:
Break deep dealloc chain in LinkedBlockingQueue with finalize method.

SynchronizedCollection holds a 'mutex' object. If you don't provide one, the default implementation uses 'this', which creates a self-referencing cycle. Not sure if this is the best way to fix the issue, but is a simple solution. Had originally marked it @Weak, but you don't really know what's being passed in. That would probably be OK too, but hard to know for sure.

kpgalligan commented:
Added a simple test. Not sure its even worth it, but feel developer guilt submitted testless code.

lukhnos commented:
When we fixed the similar issue in `LinkedList`, we added a test and identified that cut-off number [here](https://github.com/google/j2objc/blob/master/jre_emul/Tests/com/google/j2objc/LinkedListTest.java). So if you use the `SIZE_LARGE` there without the fix, you should be able to stack-overflow the test reliably.

kpgalligan commented:
Ah, cool. Will take a look.

On Tue, Oct 18, 2016 at 1:38 PM, Lukhnos Liu <notifications@github.com>
wrote:

> When we fixed the similar issue in LinkedList, we added a test and
> identified that cut-off number here
> <https://github.com/google/j2objc/blob/master/jre_emul/Tests/com/google/j2objc/LinkedListTest.java>.
> So if you use the SIZE_LARGE there without the fix, you should be able to
> stack-overflow the test reliably.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/j2objc/pull/812#issuecomment-254582597>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAELIDP0kPnfa53-S0OKo-Uvaw5lIXsfks5q1QQxgaJpZM4KaBBh>
> .
>



-- 
Kevin Galligan
http://about.me/kpgalligan

Share the Motorhead:
https://www.youtube.com/playlist?list=PL0mtSKSOY5MtzRWJnJh3FAVBR9ynCMe2W

kstanger commented on jre_emul/android/platform/libcore/ojluni/src/main/java/java/util/Collections.java:
```diff
@@ -1848,7 +1848,7 @@ public boolean equals(Object o) {
         SynchronizedCollection(Collection<E> c) {
             this.c = Objects.requireNonNull(c);
-            mutex = this;
+            mutex = new Object();
```
Nice fix. Could you please branch this file over to ojluni/src/objc. Our convention is for modified sources to be moved from src/main to src/objc. Sorry for the extra work.

kpgalligan replied:
I moved Collections. However, I don't see a similar folder structure for luni/LinkedBlockingQueue. Let me know if that also gets moved somewhere. Also moved LinkedBlockingQueueTest closer to where the LinkedListTest file is, but that may be incorrect. If there's a source org doc, let me know. Will take a look. Speaking of, if there's a "here's how we set up our build env" doc, that would be great too. My round trip is currently change file, 'make dist', and run in my downstream code.

kstanger replied:
LinkedBlockingQueue.java can stay where it is, we only started maintaining a modified vs unmodified split with the newer OpenJDK sources.

I think this would be the best info on building j2objc: http://j2objc.org/docs/Building-J2ObjC.html
A couple extra tips that I should probably add to that doc:
You can often use "make jre_emul_dist" in place of "make dist" if you don't need junit, guava, etc.
If you only need one or two architectures you can set J2OBJC_ARCHS in your environment to avoid building all 8 archs. eg. 'export J2OBJC_ARCHS="macosx iphone64"'

kstanger commented:
Thanks for the fixes, and the added test!
